### PR TITLE
Bone features

### PR DESCRIPTION
**Pull request by fire**
WIP still needs to be fixed.

Todo:

~~Animation index is always 0.~~

No matching animations

Calculating stats on the features is broken.

Cross feature normalization???

Fixes: https://github.com/GuilhermeGSousa/godot-motion-matching/issues/56

### CODE REVIEW

**Comment by GuilhermeGSousa** [comment]
Let me know if you get this to give some nice selections using the trajectory feature as well!

> Cross feature normalization???
I think what you might mean by this is feature weights. Not all features should have the same influence on selection, and right now they do. From experience, the trajectory feature usually should have a larger influence on selection than bone state.

**Comment by fire** [comment]
Need some help on this. I'm missing some theory

**Comment by fire** [comment]
```
How motion matching works by Gui!

Motion Matching is a technique that allows us to jump from any frame of any animation, to any frame of any other animation. Like an over-complicated state machine if you'd like! It works in two parts

Baking

Baking describes the process of creating a database from all our animations, that we can query at runtime based on what's going on in game, to give us the animation frame that best fits the current in-game state.

This is done by generating a big NxM matrix (what we call motion_data in MMLibrary), where
N = total number of frames on all our animations
M = sum of the dimensions of all features
Where a feature is simply something that describes whats going on. It could be the character's velocity (dimension 3), or the next 3 positions of the character in the future (dimension 3 * 3)

This results in something like this (see figure 1)
 
Runtime
Now that we have our database, we can use it to find the best animation for what's going on in game. To do this, all we need to do is to compute a vector of dimension M with the same information as a row from the matrix in figure 1. The difference is that instead of making this array from animation data, we must find a way to do it using information from the character and the level around it.

Once we have this vector, all we need to do is to compute the vector distance between it and each of the lines of the matrix. Whichever has the lowest error wins! And we then play that animation at that frame
Image
iFire — Yesterday at 01:22
how long do we play a frame?
Gui — Yesterday at 01:23
That's up to us, the "time between queries" is one of the parameters I have on the MMAnimationNode.
There are also situations that should force a query (and therefore an animation switch) to happen, like if the current animation that is playing has ended
I think right now we're doing 2 queries per second or something like that, so each animation plays for 0.5s, but it could be much longer than that
iFire — Yesterday at 01:25
Also on the web browser the camera clips through the players line of sight
I'm trying to see if we can make the web demo look good
it changes direction every 0.5 seconds
Gui — Yesterday at 01:26
I could see idle animations do queries much less often for example. This is another reason why doing this with animation node is nice, we can have different query frequencies on each node
and switch between nodes using a state machine
```

image

File: src/features/mm_bone_data_feature.cpp [hunk 1]
```diff
@@ -30,49 +30,119 @@
...
+            const Vector3 axis_x = rotation.get_column(Vector3::AXIS_X);
+            result.append(axis_x.x);
+            result.append(axis_x.y);
+            result.append(axis_x.z);
+            const Vector3 axis_y = rotation.get_column(Vector3::AXIS_Y);
+            result.append(axis_y.x);
+            result.append(axis_y.y);
+            result.append(axis_y.z);
+        } else {
+            result.append(0.0f);
```
Comment (fire) [inline src/features/mm_bone_data_feature.cpp:108]: This is an invalid identity Basis.

File: src/features/mm_bone_data_feature.cpp [hunk 2]
```diff
@@ -30,49 +30,119 @@
...
+        }
+
+        if (scale_track != -1) {
+            const Vector3 scale = p_animation->scale_track_interpolate(scale_track, time);
+            result.append(scale.x);
+            result.append(scale.y);
+            result.append(scale.z);
         } else {
             result.append(0.0f);
```
Comment (fire) [inline src/features/mm_bone_data_feature.cpp:90]: Scale tracks should be 1,1,1

**Pull request closed without merging**